Tasks/task2 completed by Кирило Юрченко

UnoMomento12 opened the pull request.

KyryloYurchenko committed:
A lot of work done

but view is not complete

KyryloYurchenko committed:
menu start

KyryloYurchenko committed:
It works! Right?

Deleted test class Entities
Implemented navigator and menu

KyryloYurchenko committed:
Code clean-up, bug fix and optimization

KyryloYurchenko committed:
Removed setters from repositories

KyryloYurchenko committed:
Improvements

1) Improved category reference restoration : If recipe was deleted from file, its reference in category will be deleted upon loading category data from file.
2) Divided method CreateAndGet(Category/Recipe) into two methods with corresponding functionality of creating and getting.

KyryloYurchenko committed:
minor cosmetic changes

1) Renamed ID property to Id.
2) Rewritten interpolated strings

KyryloYurchenko committed:
Json files

I forgot to add them

KyryloYurchenko committed:
Deleted code repeat

serhii-shkodenko commented on task2/Controllers/CategoryController.cs:
```diff
@@ -0,0 +1,79 @@
+        public Category CreateAndGetCategory(Category category)
+        {
+            CreateCategory(category);
+            return category;
```
Вы хотите просигнализировать пользователю, что операция по добавлению категории была успешно завершена. Например, этой цели может служить простое булевское значение, количество созданных записей в источнике данных, либо объект, возвращенный из источника данных. У Вас же возвращается то же, что и приходит.

serhii-shkodenko commented on task2/Controllers/CategoryController.cs:
```diff
@@ -0,0 +1,79 @@
+            _unitOfWork.Save();
+        }
+
+        //public void Add(Category category)
```
Если этот код более не нужен, удалите его.

serhii-shkodenko commented on task2/Controllers/IngredientController.cs:
```diff
@@ -0,0 +1,33 @@
+        public IngredientController(UnitOfWork unitOfWork) : base(unitOfWork)
+        {
+        }
+        //public Ingredient CreateAndGetIngredient(string ingName , string denomination)
```
Если этот код более не нужен, удалите его.

serhii-shkodenko commented on task2/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,93 @@
+            _unitOfWork.Save();
+        }
+
+        //public void AddSteps(Recipe recipe, List<string> steps)
```
Если этот код более не нужен, удалите его.

serhii-shkodenko commented on task2/Models/BaseModel.cs:
```diff
@@ -0,0 +1,11 @@
+        public string Id { get; set; } = default;
+        public string Name { get; set; } = default;
+
+        public BaseModel() { }
```
Зачем здесь дефолтный конструктор?

serhii-shkodenko commented on task2/Models/Category.cs:
```diff
@@ -0,0 +1,17 @@
+    class Category : BaseModel
+    {
+
+        public string ParentId { get; set; } = null;
```
Для типа стринг дефолтное значение и так `null`.

serhii-shkodenko commented on task2/Models/Category.cs:
```diff
@@ -0,0 +1,17 @@
+        public List<string> RecipeIds { get; set; } = new List<string>();
+
+        [JsonIgnore]
+        public Category Parent { get; set; } = null;
```
Для типа стринг дефолтное значение и так `null`.

serhii-shkodenko commented on task2/Models/Ingredient.cs:
```diff
@@ -0,0 +1,8 @@
+{
+    class Ingredient : BaseModel
+    {
+        public string Denomination { get; set; } = "";
```
Зачем инициалазируете как пустую строку?

serhii-shkodenko commented on task2/Models/IngredientDetail.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public Ingredient Ingredient { get; set; }
+        public double Amount { get; set; }
+        public IngredientDetail(Ingredient ing, double amount)
```
Обычно наличие подобного конструктора для инициализации всех свойств подразумевает, что у них не будет сеттеров. Рассмотрите целесобразность этого конструктора.

serhii-shkodenko commented on task2/Navigator.cs:
```diff
@@ -0,0 +1,158 @@
+        {
+            _unitOfWork = unit;
+            _current = null;
+            _root = null;
```
Значения `_current` и `_root`и так null.

KyryloYurchenko committed:
Changes applied

serhii-shkodenko commented on task2/Models/Category.cs:
```diff
@@ -0,0 +1,17 @@
+        public List<string> RecipeIds { get; set; } = new List<string>();
+
+        [JsonIgnore]
+        public Category Parent { get; set; } = null;
```
Значение по умолчанию для рефренсных типов и так `null`. Не нужно это свойство таким образом инициалазировать.

serhii-shkodenko commented on task2/Models/Category.cs:
```diff
@@ -0,0 +1,17 @@
+    class Category : BaseModel
+    {
+
+        public string ParentId { get; set; } = default;
```
Дефолтное значение для типа string `null`. Неясно, какую Вы цель преследуете.

serhii-shkodenko commented on task2/Models/Recipe.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    class Recipe : BaseModel
+    {
+        public string Description { get; set; } = default;
```
Дефолтное значение для типа string `null`. Неясно, какую Вы цель преследуете.

serhii-shkodenko commented on task2/Navigator.cs:
```diff
@@ -0,0 +1,156 @@
+            if (retr is Category)
+            {
+                _root = _current;
+                _current = retr as Category;
```
Вверху Вы уже знаете, что работает с категорией. Подумайте, как сократить количество кастов до одного.

serhii-shkodenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+                string input = Console.ReadLine().ToLower().Trim();
+                if (input == "exit") break;
+                int option;
+                bool isNumber = Int32.TryParse(input, out option);
```
Воспользуйтесь псевдонимом для типа.

serhii-shkodenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+            if(choice == false)
+            {
+                Console.Write($"Please enter an id of a category where created {item} will be stored:");
+                while (Int32.TryParse(Console.ReadLine().Trim(), out categoryID) == false || navig.GetCategory(categoryID) == null)
```
Воспользуйтесь псевдонимом для типа. Общепринятым является использование именно псевдонимов, а не полных имен.

serhii-shkodenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,86 @@
+            {
+                cat.Recipes = RestoreRecipesInCategory(cat);
+                cat.Parent = RestoreParent(cat);
+            }
```
Зачем Вы храните данные? Я бы не усложнял так слой для доступа к данным, пусть он предоставляет данные по запросу. Задачу хранения можно возложить на другие слои.

UnoMomento12 replied:
Вот тут не понял, что имеется в виду? Данные у меня считываются из файла, эти методы восстанавливают(присваивают) референсы на объекты, которые категория должна содержать.

UnoMomento12 replied:
Если же это относилось к строке ниже где я опять информацию сохраняю, то я добавил её на такой кейс: есть 3 файла, и в файле с рецептами я удоляю один рецепт, в файле категории записан Id рецепта, при запуске программы, если при восстановлении референсов рецептов его не находит, id рецепта в списке рецептов категории удаляется, изменения сохраняются сразу( на тот случай если никаких действий в работающей программе совершено не будет). Там только можно изменить из общего сейва всех файлов на именно сейв файла категории.

serhii-shkodenko replied:
Вот эту логику наполнения рецептов ингридиентами и наполнения категории рецептами лучше разместить в методах соответствующих репозиториев. В общем случае конструкторы служат цели дать потребителю готовый для работы объект, логика с запросами здесь смотрится неуместной.

UnoMomento12 replied:
Я думал об этом, но тогда получается, что бы восстановить ингредиенты в рецептах, прийдется передавать репозиторий игредиентов в репозиторий рецептов, а сам репозиторий рецептов в репозиторий категорий.

serhii-shkodenko replied:
Да, скорее всего, Вы правы. Пусть реализация пока остается такой. Посмотрим, что скажет @alex-geraschenko. Но интерейс все-таки выделите. Он в дальнейшем Вам пригодится.

serhii-shkodenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,86 @@
+
+namespace task2.UnitsOfWork
+{
+    class UnitOfWork 
```
В классической реализации паттерна UnitOfWork, этот класс должен реализовывать интерфейс IUnitOfWork.

KyryloYurchenko committed:
Changes applied

Осталось решить что делать с восстановлением референсов.

alex-geraschenko commented on task2/Controllers/BaseController.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    abstract class BaseController : IController
+    {
+        protected IUnitOfWork _unitOfWork;
```
Именование с подчеркиванием вначале принято только для приватных полей. Для `protected` лучше использовать `PascalCase`

alex-geraschenko commented on task2/Category.json:
```diff
@@ -0,0 +1,69 @@
+﻿[
+  {
+    "ParentId": null,
+    "RecipeIds": [
```
Лучше сделать обратную зависимость. Не категория знает о рецептах, а рецепт знает, к какой категории он относится

alex-geraschenko commented on task2/Controllers/CategoryController.cs:
```diff
@@ -0,0 +1,52 @@
+            var item = _unitOfWork.Categories.SingleOrDefault(x => string.Equals(x.Name, category.Name, StringComparison.OrdinalIgnoreCase) && x.ParentId == category.ParentId);
+            if (item != null)
+            {
+                category = item;
```
Плохая практика перезаписывать входные параметры

alex-geraschenko commented on task2/Controllers/IController.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace task2.Controllers
+{
+    interface IController
+    {
```
Зачем нужен пустой интерфейс?

alex-geraschenko commented on task2/Controllers/RecipeController.cs:
```diff
@@ -0,0 +1,61 @@
+            if (checker != null)
+            {
+                Console.WriteLine($"Recipe {checker.Name} : {checker.Id} already exists");
+                recipe = checker;
```
Плохая практика перезаписывать входные параметры

alex-geraschenko commented on task2/Models/Category.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+
+        public string ParentId { get; set; } 
+        public List<string> RecipeIds { get; set; } = new List<string>();
```
Лучше сделать обратную зависимость. Не категория знает о рецептах, а рецепт знает, к какой категории он относится

alex-geraschenko commented on task2/Navigator.cs:
```diff
@@ -0,0 +1,156 @@
+            Console.WriteLine($"Subcategories and recipes in {_current.Name} category.");
+            WriteRootNavigator();
+            Console.WriteLine("Recipes:");
+            for( int b = _recipesStart; b < _subItems?.Count; b++) 
```
Проверка на `null` тут не имеет смысла

alex-geraschenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+                while (true)
+                {
+                    Console.Write("Amount:");
+                    parsed = Double.TryParse(Console.ReadLine(), out amount);
```
Воспользуйтесь псевдонимом для типа.

alex-geraschenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+                    {
+                        break;
+                    } else
+                    {
```
Зачем нужен пустой `else`?

alex-geraschenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+                if (step == "-1") break;
+                steps.Add(step);
+            }
+            return new Recipe() { Name = name, Description = description, Steps = steps };
```
Это ответсвтенность контроллера создавать рецепт

alex-geraschenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+                string answer;
+                while (true)
+                {
+                    answer = Console.ReadLine().Trim().ToLower();
```
Попробуйте `System.ConsoleKey`

alex-geraschenko commented on task2/Program.cs:
```diff
@@ -0,0 +1,215 @@
+                    }
+                }
+            }
+            if(choice == false)
```
Выражение можно упростить

alex-geraschenko commented on task2/Repositories/Repository.cs:
```diff
@@ -0,0 +1,52 @@
+    abstract class Repository<T> : IRepository<T> where T: BaseModel, new()
+    {
+        private IDataManager _dataManager;
+        protected List<T> _items;
```
Именование с подчеркиванием вначале принято только для приватных полей. Для `protected` лучше использовать PascalCase

alex-geraschenko commented on task2/Repositories/IRepository.cs:
```diff
@@ -0,0 +1,18 @@
+        void Save();
+        T Get(string guid);
+        IEnumerable<T> GetItems();
+        IEnumerable<T> Where(Func<T, bool> predicate);
```
Метод `Where` избыточен

alex-geraschenko commented on task2/Repositories/IRepository.cs:
```diff
@@ -0,0 +1,18 @@
+        T Get(string guid);
+        IEnumerable<T> GetItems();
+        IEnumerable<T> Where(Func<T, bool> predicate);
+        T SingleOrDefault(Func<T, bool> predicate);
```
Метод `SingleOrDefault` избыточен

UnoMomento12 replied:
@alex-geraschenko Я понять не могу, в чем избыточность этих двух методов? Я их использую 8 и 9 раз соответственно.

alex-geraschenko replied:
Они дублируют то, что есть в `System.Linq`.

UnoMomento12 replied:
@alex-geraschenko Ну так смысл в том, что бы в контроллерах вызывать этот метод через UnitOfWork.RepositoryName.SingleOrDefault(), а не UnitOfWork.RepositoryName.ItemsInRepository.SingleOrDefault() . 
Или я не правильно понимаю что именно дублируется?
ItemsInRepository это protected List

alex-geraschenko replied:
Не `UnitOfWork.RepositoryName.ItemsInRepository.SingleOrDefault()`, а `UnitOfWork.RepositoryName.GetIAll().SingleOrDefault()`

Классическая реализация `Repository` включает в себя только 5 методов. Все остальное происходит или в сервисах или в контроллерах или при помощи шаблона `Specification`

UnoMomento12 replied:
Понятно, спасибо за разъяснение.

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+            RestoreIngredientsInRecipes();
+            RestoreRecipesInCategories();
+
+            if (_refDeleted) Categories.Save();
```
В конструкторе не должно быть логики сохранения. Это костыль

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+                    result.Add(a);
+                }
+            }
+            if (_refDeleted) { DeleteDeadReferences(category, toDeleteInCategory); }
```
Нарушение принципа **Single Responsibility**

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+            Categories.Save();
+        }
+
+        private void RestoreIngredientsInRecipes()
```
Эту логику лучше вынести из `UnitOfWork` в контроллер

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+                recipe.Ingredients = RestoreIngredients(recipe);
+            }
+        }
+        private void RestoreRecipesInCategories()
```
Эту логику лучше вынести из `UnitOfWork` в контроллер

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+            }
+        }
+        
+        private List<IngredientDetail> RestoreIngredients(Recipe recipe)
```
Эту логику лучше вынести из `UnitOfWork` в контроллер

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+            }
+            return result;
+        }
+        private Category RestoreParent(Category category)
```
Эту логику лучше вынести из `UnitOfWork` в контроллер

alex-geraschenko commented on task2/UnitsOfWork/UnitOfWork.cs:
```diff
@@ -0,0 +1,94 @@
+        {
+            return Categories.SingleOrDefault(x=> x.Id == category.ParentId);
+        }
+        private List<Recipe> RestoreRecipesInCategory(Category category)
```
Эту логику лучше вынести из `UnitOfWork` в контроллер

KyryloYurchenko committed:
I hope it's The last

KyryloYurchenko committed:
minor code-repeat fix

KyryloYurchenko committed:
Made method name more explicit

KyryloYurchenko committed:
Excessive methods were removed

KyryloYurchenko committed:
Adjustments

1) Реализован контекст данных (класс Entities)                                                    2) Следуя примеру из https://docs.microsoft.com/ru-ru/aspnet/mvc/overview/older-versions/getting-started-with-ef-5-using-mvc-4/implementing-the-repository-and-unit-of-work-patterns-in-an-asp-net-mvc-application
репозитории в UnitOfWork  не имеют собственной реализации сохранения данных, сохранение идет из констекста ( точнее вызов идет из UnitOfWork)

KyryloYurchenko committed:
Fix of Dependency Inversion violation

alex-geraschenko commented on task2/EntityList;/Entities.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System.Collections.Generic;
```
Опечатка в пути к файлу

UnoMomento12 replied:
@alex-geraschenko Исправил

KyryloYurchenko committed:
Fixed a folder name typo

alex-geraschenko reviewed (approved):
Хорошо, можно мержить
